File: frontend/src/api/design.ts
import { api, apiJson } from './client'
import type { DesignJob, DesignJobListOut } from './types'

export type DesignCreateBody = {
  name?: string | null
  fold_job_id?: string | null
  designed_chains?: string
  num_seq_per_target?: number
  sampling_temp?: number
  seed?: number
  backbone_noise?: number
  omit_aas?: string
}

export async function fetchDesignJobs(limit = 50) {
  return apiJson<DesignJobListOut>(`/api/design-jobs?limit=${limit}`)
}

export async function fetchDesignJob(id: string) {
  return apiJson<DesignJob>(`/api/design-jobs/${id}`)
}

export async function createDesignJob(body: DesignCreateBody) {
  return apiJson<DesignJob>('/api/design-jobs', { method: 'POST', data: body })
}

export async function uploadDesignJob(structure: File, body: DesignCreateBody) {
  const fd = new FormData()
  fd.append('structure', structure)
  if (body.name) fd.append('name', body.name)
  fd.append('designed_chains', body.designed_chains ?? '')
  fd.append('num_seq_per_target', String(body.num_seq_per_target ?? 8))
  fd.append('sampling_temp', String(body.sampling_temp ?? 0.1))
  fd.append('seed', String(body.seed ?? 0))
  fd.append('backbone_noise', String(body.backbone_noise ?? 0))
  fd.append('omit_aas', body.omit_aas ?? 'X')
  return apiJson<DesignJob>('/api/design-jobs/upload', { method: 'POST', data: fd })
}

export async function deleteDesignJob(id: string) {
  await apiJson(`/api/design-jobs/${id}`, { method: 'DELETE' })
}

export async function downloadDesignFile(id: string, filename: string) {
  const response = await api.get(`/api/design-jobs/${id}/files/${encodeURIComponent(filename)}`, {
    responseType: 'blob',
  })
  const url = URL.createObjectURL(response.data)
  const anchor = document.createElement('a')
  anchor.href = url
  anchor.download = filename
  anchor.style.display = 'none'
  document.body.appendChild(anchor)
  anchor.click()
  window.setTimeout(() => {
    URL.revokeObjectURL(url)
    anchor.remove()
  }, 1000)
}

File: app/schemas/design.py
"""Protein design schemas."""

from pydantic import BaseModel, Field

from .common.jobs import JobOut


class DesignJobCreate(BaseModel):
    name: str | None = Field(default=None, max_length=128)
    fold_job_id: str | None = None
    designed_chains: str = Field(default="", max_length=64, description="要设计的链，空格分隔，如 H A；空=全部")
    num_seq_per_target: int = Field(default=8, ge=1, le=64)
    sampling_temp: float = Field(default=0.1, ge=0.05, le=1.0)
    seed: int = Field(default=0, ge=0, le=999999)
    backbone_noise: float = Field(default=0.0, ge=0.0, le=1.0)
    omit_aas: str = Field(default="X", max_length=32)


class DesignJobOut(JobOut):
    pass


class DesignJobListOut(BaseModel):
    items: list[DesignJobOut]
    total: int
